In [4]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [5]:
# Data for Tesla, Apple, and Microsoft for the last three fiscal years
data = {
    "Company": ["Tesla", "Tesla", "Tesla", "Apple", "Apple", "Apple", "Microsoft", "Microsoft", "Microsoft"],
    "Year": [2021, 2022, 2023, 2021, 2022, 2023, 2021, 2022, 2023],
    "Total Revenue ($ millions)": [53823, 81462, 96773, 365817, 394328, 383934, 168088, 198270, 211915],
    "Net Income ($ millions)": [5644, 12587, 14974, 94680, 99803, 96995, 61271, 72738, 70583],
    "Total Assets ($ millions)": [62131, 82338, 106618, 351002, 352755, 352583, 333779, 364840, 411179],
    "Total Liabilities ($ millions)": [30548, 36440, 43009, 287912, 302083, 290437, 191791, 198232, 226452],
    "Cash Flow from Operating Activities ($ millions)": [11497, 14724, 13256, 104038, 122151, 110543, 76742, 89041, 87648]
}

In [6]:
# Create a DataFrame
df = pd.DataFrame(data)

In [8]:
# Save the DataFrame to an Excel file
file_path = '/content/financial_data_analysis.xlsx'
df.to_excel(file_path, index=False)

# Displaying the file to the user
file_path

'/content/financial_data_analysis.xlsx'

In [9]:
# Proceeding with the analysis using pandas for trends and insights

# Let's reload the data for analysis and calculate trends such as year-over-year growth for each company.
df = pd.DataFrame(data)

# Calculate the percentage change year-over-year for each financial metric by company
df['Revenue Growth (%)'] = df.groupby('Company')['Total Revenue ($ millions)'].pct_change() * 100
df['Net Income Growth (%)'] = df.groupby('Company')['Net Income ($ millions)'].pct_change() * 100
df['Asset Growth (%)'] = df.groupby('Company')['Total Assets ($ millions)'].pct_change() * 100
df['Liability Growth (%)'] = df.groupby('Company')['Total Liabilities ($ millions)'].pct_change() * 100
df['Cash Flow Growth (%)'] = df.groupby('Company')['Cash Flow from Operating Activities ($ millions)'].pct_change() * 100

# Display the updated dataframe with growth percentages
df.head()


Company  Year  Total Revenue ($ millions)  Net Income ($ millions)  \
0   Tesla  2021                       53823                     5644   
1   Tesla  2022                       81462                    12587   
2   Tesla  2023                       96773                    14974   
3   Apple  2021                      365817                    94680   
4   Apple  2022                      394328                    99803   

   Total Assets ($ millions)  Total Liabilities ($ millions)  \
0                      62131                           30548   
1                      82338                           36440   
2                     106618                           43009   
3                     351002                          287912   
4                     352755                          302083   

   Cash Flow from Operating Activities ($ millions)  Revenue Growth (%)  \
0                                             11497                 NaN   
1                                             14724           51.351653   
2                                             13256           18.795267   
3                                            104038                 NaN   
4                                            122151            7.793788   

   Net Income Growth (%)  Asset Growth (%)  Liability Growth (%)  \
0                    NaN               NaN                   NaN   
1             123.015592         32.523217             19.287678   
2              18.964010         29.488207             18.026894   
3                    NaN               NaN                   NaN   
4               5.410858          0.499427              4.921990   

   Cash Flow Growth (%)  
0                   NaN  
1             28.068192  
2             -9.970117  
3                   NaN  
4             17.409985

# **SUMMARY**


---


### **Tesla Trends:**
1. **Revenue Growth**:
   - Tesla's revenue increased significantly year-over-year, with a 51.35% growth from 2021 to 2022 and 18.80% from 2022 to 2023.
   - This indicates strong growth, likely driven by increased vehicle deliveries and expanding production capacity.

2. **Net Income Growth**:
   - Tesla’s net income growth was highly impressive, with a 123.02% increase from 2021 to 2022 and an 18.96% increase from 2022 to 2023.
   - This showcases improved profitability, reflecting cost controls, greater economies of scale, and higher margins.

3. **Asset Growth**:
   - Tesla's total assets grew by 32.52% from 2021 to 2022 and 29.49% from 2022 to 2023.
   - This highlights Tesla’s increasing investments in infrastructure, production facilities, and overall business expansion.

4. **Liability Growth**:
   - Total liabilities increased by 19.29% from 2021 to 2022 and 18.03% from 2022 to 2023.
   - This reflects Tesla’s growing financial obligations, including debt financing for expansion.

5. **Cash Flow from Operating Activities**:
   - Tesla experienced strong growth in cash flow in 2022, up 28.07%, but saw a slight decline of 9.97% in 2023.


---


### **Apple Trends:**
1. **Revenue Growth**:
   - Apple saw a modest 7.79% revenue growth from 2021 to 2022 but experienced a slight decline of 2.64% from 2022 to 2023.
   - This may indicate market saturation for its flagship products or macroeconomic challenges affecting sales.

2. **Net Income Growth**:
   - Apple’s net income grew by 5.41% from 2021 to 2022 but decreased by 2.81% in 2023.
   - Similar to revenue, Apple’s profitability could be impacted by macroeconomic factors or shifting consumer behavior.

3. **Asset Growth**:
   - Apple’s assets remained relatively stable, with a minimal increase of 0.50% from 2021 to 2022 and a slight decrease of 0.05% in 2023.

4. **Liability Growth**:
   - Total liabilities grew by 4.92% in 2022 but decreased by 3.86% in 2023.
   - This could indicate that Apple is paying off debt or reducing financial obligations as part of its financial strategy.

5. **Cash Flow from Operating Activities**:
   - Apple saw a solid increase in cash flow in 2022, up 17.41%, though there was a decline of 9.48% in 2023.


---

### **Microsoft Trends:**
1. **Revenue Growth**:
   - Microsoft experienced steady revenue growth, with an 18% increase from 2021 to 2022 and a 6.9% increase from 2022 to 2023.
   - This indicates strong and consistent performance, likely driven by growth in cloud services, enterprise software, and hardware sales.

2. **Net Income Growth**:
   - Microsoft’s net income increased by 18.63% from 2021 to 2022 but decreased by 2.96% from 2022 to 2023.
   - The drop in 2023 net income might be attributed to rising operational costs or specific market factors, but overall profitability remains strong.

3. **Asset Growth**:
   - Microsoft’s total assets grew by 9.3% from 2021 to 2022 and 12.7% from 2022 to 2023.
   - This steady growth reflects Microsoft's continued investments in cloud infrastructure, acquisitions, and innovation.

4. **Liability Growth**:
   - Liabilities grew by 3.36% from 2021 to 2022 and a larger 14.2% from 2022 to 2023.
   - This suggests Microsoft may have taken on additional debt or increased its obligations to fund growth and innovation.

5. **Cash Flow from Operating Activities**:
   - Microsoft’s cash flow from operations increased by 16.09% in 2022 but decreased slightly by 1.57% in 2023.
   - Despite the slight dip, Microsoft continues to generate robust cash flow, which is essential for its R&D, acquisitions, and dividend payouts.

---

### **Overall Insights Across Tesla, Apple, and Microsoft**:
1. **Growth Leaders**:
   - **Tesla**: Rapid revenue and income growth driven by increased production and sales.
   - **Microsoft**: Steady revenue growth, strong cash flow, and leadership in cloud services.
   - **Apple**: Consistent but slowing growth, facing challenges in maintaining revenue growth at the same level.

2. **Profitability**:
   - Tesla's profitability is improving rapidly, while Apple and Microsoft have consistently high profit margins but saw slight declines in 2023.

3. **Investment Insight for Chatbot**:
   - **Tesla** could be positioned as a high-growth investment for users seeking rapid gains in the tech and electric vehicle sectors.
   - **Microsoft** could be emphasized for its consistent performance and leadership in cloud computing, offering stability.
   - **Apple** remains a solid investment with steady cash flow and profitability, though recent trends suggest the need for innovation to spur growth.



---